src/compiler/parser.ts

```diff
@@ -3646,34 +3649,72 @@ module ts {
             }
         }
 
-        function isStartOfStatement(inErrorRecovery: boolean): boolean {
-            // Functions, variable statements and classes are allowed as a statement.  But as per
-            // the grammar, they also allow modifiers.  So we have to check for those statements 
-            // that might be following modifiers.This ensures that things work properly when
-            // incrementally parsing as the parser will produce the same FunctionDeclaraiton, 
-            // VariableStatement or ClassDeclaration, if it has the same text regardless of whether 
-            // it is inside a block or not.
-            if (isModifier(token)) {
-                let result = lookAhead(parseVariableStatementOrFunctionDeclarationOrClassDeclarationWithDecoratorsOrModifiers);
-                if (result) {
-                    return true;
+        function isIdentifierOrKeyword() {
+            return token >= SyntaxKind.Identifier;
+        }
+
+        function nextTokenIsIdentifierOrKeywordOnSameLine() {
+            nextToken();
+            return isIdentifierOrKeyword() && !scanner.hasPrecedingLineBreak();
+        }
+
+        function parseDeclarationFlags(): StatementFlags {
+            while (true) {
+                switch (token) {
+                    case SyntaxKind.VarKeyword:
+                    case SyntaxKind.LetKeyword:
+                    case SyntaxKind.ConstKeyword:
+                    case SyntaxKind.FunctionKeyword:
+                    case SyntaxKind.ClassKeyword:
+                    case SyntaxKind.EnumKeyword:
+                        return StatementFlags.Statement;
+                    case SyntaxKind.InterfaceKeyword:
+                    case SyntaxKind.TypeKeyword:
+                        nextToken();
+                        return isIdentifierOrKeyword() ? StatementFlags.Statement : StatementFlags.None;
+                    case SyntaxKind.ModuleKeyword:
+                    case SyntaxKind.NamespaceKeyword:
+                        nextToken();
+                        return isIdentifierOrKeyword() || token === SyntaxKind.StringLiteral ? StatementFlags.ModuleElement : StatementFlags.None;

```

Comment:
In general, even for the trivial example above, I find the second (parenthesized) expression far clearer.  Parenthesis remove any assumptions and ambiguity from both the authors intent, and the readers comprehension.

That said, this is probably more subjective than absolute (as much of the feedback has been) and we should probably agree as a team on a coding guideline to be consistent and avoid future philosophical debates.  For example, the "C# Coding Conventions"  (https://msdn.microsoft.com/en-us/library/ff926074.aspx) state:

> Use parentheses to make clauses in an expression apparent, as shown in the following code. `if ((val1 > val2) && (val1 > val3))`

.. and the Java guidelines (http://www.oracle.com/technetwork/java/codeconventions-150003.pdf) state ..

> It is generally a good idea to use parentheses liberally in expressions involving mixed operators to avoid operator precedence problems. Even if the operator precedence seems clear to you, it might not be to others—you shouldn’t assume that other programmers know precedence as well as you do.

```java
if (a == b && c == d)     // AVOID!
if ((a == b) && (c == d)) // RIGHT
```

Yet some of the Google coding guidelines (https://google-styleguide.googlecode.com/svn/trunk/javascriptguide.xml#Parentheses) state of parenthesis:

> Use sparingly and in general only where required by the syntax and semantics.

We should just pick one and make it policy for the codebase.